core/trino-main/src/main/java/io/trino/execution/buffer/SpoolingExchangeOutputBuffer.java

```diff
@@ -166,34 +170,43 @@ public void enqueue(int partition, List<Slice> pages)
     @Override
     public void setNoMorePages()
     {
-        stateMachine.noMorePages();
-        destroy();
+        if (!stateMachine.noMorePages()) {
+            return;
+        }
+
+        exchangeSink.finish().whenComplete((value, failure) -> {
+            if (failure != null) {
+                stateMachine.fail(failure);
+            }
+            else {
+                stateMachine.finish();
+            }
+            updateMemoryUsage(0);
+        });
     }
 
     @Override
     public void destroy()
     {
-        if (stateMachine.finish()) {
-            try {
-                exchangeSink.finish();
-            }
-            finally {
-                updateMemoryUsage(exchangeSink.getMemoryUsage());
-            }
-        }
+        // Abort the buffer if it hasn't been finished. This is possible when a task is cancelled early by the coordinator.
+        // Task cancellation is not supported (and not expected to be requested by the scheduler when the spooling exchange
+        // is in use) as the task output is expected to be deterministic.
+        abort();
     }
 
     @Override
     public void abort()
     {
-        if (stateMachine.abort()) {
-            try {
-                exchangeSink.abort();
-            }
-            finally {
-                updateMemoryUsage(0);
-            }
+        if (!stateMachine.abort()) {
+            return;
         }
+
+        exchangeSink.abort().whenComplete((value, failure) -> {
+            if (failure != null) {
+                log.warn(failure, "Error aborting exchange sink");
```

Comment:
why not `error`?

Reply:
It looks like generally we are using `warn` to log errors that do not impact query success. For example:

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/spiller/FileSingleStreamSpillerFactory.java#L160
https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/sql/planner/SplitSourceFactory.java#L124
https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePageSink.java#L245